File: 프로그래머스/단어 변환.js
// [프로그래머스] 단어 변환

function solution(begin, target, words) {
    if (words.indexOf(target) === -1)
        return 0;
    let visited = new Array(words.length).fill(false);
    let queue = [[begin, 0]];
    let value;
    while (queue.length > 0) {
        value = queue.shift();

        for (let i = 0; i < words.length; i++) {
            if (visited[i] === false) {
                let diff = 0;
                let index = words.indexOf(words[i]);

                for (let k = 0; k < words[0].length; k++) {
                    if (value[0][k] !== words[i][k])
                        diff++;
                }
                if (diff === 1 && visited[index] === false) {
                    if (words[i] === target) return value[1] + 1;
                    queue.push([words[i], value[1] + 1]);
                    visited[index] = true;
                };
            }
        }
    }
    return 0;
}

File: 프로그래머스/단어 변환.py
from collections import deque


def search(begin, target, words):
    answer = 0

    queue = deque([(begin, 0)])
    visited = [False] * len(words)

    while queue:
        cur_word, count = queue.popleft()
        for i, next_word in enumerate(words):
            if not visited[i] and can_change(cur_word, next_word):
                visited[i] = True
                queue.append((next_word, count + 1))

                if next_word == target:
                    return count + 1

    return answer


def can_change(cur_word, next_word):
    diff = 0

    for i in range(len(cur_word)):
        if cur_word[i] != next_word[i]:
            diff += 1
        if diff > 1:
            return False

    return True


def solution(begin, target, words):
    answer = 0

    if target in words:
        answer = search(begin, target, words)

    return answer


print(solution("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"]))
